File: 0020-valid-parentheses/0020-valid-parentheses.ts
function isValid(s: string): boolean {
    if (s.length & 1) return false
    
    const map = new Map<string, string>([
        [')', '('],
        ['}', '{'],
        [']', '['],
    ]);
    const open = new Set(map.values());
    const close = new Set(map.keys());

    const stack: string[] = [];

    for (const c of s) {
        if (open.has(c)) {
            stack.push(c);
            continue;
        }
        if (stack.length === 0 || (stack.pop() !== map.get(c))) {
            return false;
        }
    }
    return stack.length === 0;
};

File: 0020-valid-parentheses/0020-valid-parentheses.go
type Stack []rune

func (s *Stack) IsEmpty() bool {
    return len(*s) == 0
}

func (s *Stack) Pop() rune {
    if s.IsEmpty() {
        var r rune
        return r
    }
    top := len(*s) - 1
    item := (*s)[top]
    *s = (*s)[:top]
    return item
}

func (s *Stack) Push(item rune) {
    *s = append(*s, item)
}


func isValid(s string) bool {
    if (len(s) & 1) == 1 {
        return false
    }

    bracketPairs := map[rune]rune {
        ']': '[',
        '}': '{',
        ')': '(',
    }
    
    stack := Stack{}
    
    for _, c := range s {
        if c == '[' || c == '{' || c == '(' {
            stack.Push(c)
            continue
        }
        if stack.Pop() != bracketPairs[c] {
            return false
        }
    }
    
    return stack.IsEmpty()
}